pallets/oracle/src/lib.rs

```diff
@@ -0,0 +1,511 @@
+		#[pallet::weight(10000)]
+		pub fn add_key(origin: OriginFor<T>, public: ed25519::Public) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("New key: {:?}", public);
+
+			<FragKeys<T>>::mutate(|validators| {
+				validators.insert(public);
+			});
+
+			Ok(())
+		}
+
+		/// Remove a Clamor Account ID from `FragKeys`
+
+		/// NOTE: Only the Root User of the Clamor Blockchain (i.e the local node itself) can call this function
+		#[pallet::weight(10000)] // TODO
+		pub fn del_key(origin: OriginFor<T>, public: ed25519::Public) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("Removed key: {:?}", public);
+
+			<FragKeys<T>>::mutate(|validators| {
+				validators.remove(&public);
+			});
+
+			Ok(())
+		}
+
+		/// This function stores the new price received from the Oracle into a structure that contains a list of the last N-prices.
+		#[pallet::weight(10000)] // TODO
+		pub fn store_price(
+			origin: OriginFor<T>,
+			oracle_price: OraclePrice<T::Public, T::BlockNumber>,
+			_signature: T::Signature,
+		) -> DispatchResult {
+			// This ensures that the function can only be called via unsigned transaction.
+			ensure_none(origin)?;
+
+			log::debug!("Store price: {:?}", oracle_price);
+
+			let data_hash: H256 = oracle_price.using_encoded(blake2_256).into();
+
+			let latest_price: u32 =
+				oracle_price.price.try_into().map_err(|_| Error::<T>::SystematicFailure)?;
+			ensure!(!latest_price.is_zero(), Error::<T>::PriceIsZero);
+
+			let block_number = oracle_price.block_number;
+
+			// voting
+			let threshold = T::Threshold::get();
+			if threshold > 1 {
+				// 问Gio
+				let current_votes = <EVMLinkVoting<T>>::get(&data_hash);
+				if let Some(current_votes) = current_votes {
+					// Number of votes for the key `data_hash` in EVMLinkVoting
+					if current_votes + 1u64 < threshold {
+						// Current Votes has not passed the threshold
+						<EVMLinkVoting<T>>::insert(&data_hash, current_votes + 1);
+						return Ok(())
+					} else {
+						// Current votes passes the threshold, let's remove the record
+						<EVMLinkVoting<T>>::remove(&data_hash);
+					}
+				} else {
+					// If key `data_hash` doesn't exist in EVMLinkVoting
+					<EVMLinkVoting<T>>::insert(&data_hash, 1);
+					return Ok(())
+				}
+			}
+
+			log::trace!("Store new price: {}", latest_price);
+
+			<Price<T>>::put(latest_price);
+
+			Self::deposit_event(Event::NewPrice { price: latest_price, block_number });
+			Ok(())
+		}
+
+		/// Circuit breaker!
+		/// Stop the Oracle by changing the value of **OracleStop** flag.
+		/// **true** = oracle stop from the next block,
+		/// **false** = oracle keeps running (default).
+		#[pallet::weight(10000)]
+		pub fn stop_oracle(origin: OriginFor<T>, stop_it: bool) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("Oracle stop flag: {:?}", stop_it);
+
+			IsOracleStopped::<T>::put(stop_it);
```

Comment:
shouldn't this be `set`?

Reply:
No, `put` is correct:

[[doc](https://paritytech.github.io/substrate/master/frame_support/storage/trait.StorageValue.html#tymethod.put)]
> **fn put**<Arg: EncodeLike<T>>(val: Arg)
> Store a value under this key into the provided storage instance.

> **fn set**(val: Self::Query)
> Store a value under this key into the provided storage instance; this uses the query type rather than the underlying value.

Also in Substrate only `put` is used with `StorageValue`.

Reply:
what's the difference?

Reply:
I guess the difference is in the type <Arg: EncodeLike>, but then there is no actual difference in purpose. They both achieve the same purpose. `put` seems to be the only one used, though. I can't find examples of usage of `set`.

Reply:
I can read that type difference, I was wondering what happens inside

Reply:
Here the difference:
```rust
fn put<Arg: EncodeLike<T>>(val: Arg) {
      unhashed::put(&Self::storage_value_final_key(), &val)
}
```
`set `uses `put` when there is an actual value in input, while it clear the key in case there is `None`.
```rust
fn set(maybe_val: Self::Query) {
	if let Some(val) = G::from_query_to_optional_value(maybe_val) {
		unhashed::put(&Self::storage_value_final_key(), &val)
	} else {
		unhashed::kill(&Self::storage_value_final_key())
	}
}
```

Reply:
Interesting, good to know that clears storage.